src/Model/Behavior/NullableBehavior.php

```diff
@@ -36,6 +36,10 @@ protected function _process($data, Table $table) {
 			$associations[$association->getProperty()] = $association->getName();
 		}
 
+		if ($data === null) {
```

Comment:
According to docblock thats not possible.
Can you show me an example? Or add a test case?

Reply:
I'm using https://github.com/josegonzalez/cakephp-version plugin and that seems to be sending the null data. I've been using the Shim plugin before I installed the version plugin. I'll try to get a stack trace of the issue happening. But I'm sure if you add that plugin and you'll experience the error. It's gotta be how it marshalls the data to build a version.

Reply:
A stack trace would be helpful.

Reply:
```
  | Shim\Model\Behavior\NullableBehavior::_process() - ROOT/vendor/dereuromark/cakephp-shim/src/Model/Behavior/NullableBehavior.php, line 39
-- | --
  | Shim\Model\Behavior\NullableBehavior::_process() - ROOT/vendor/dereuromark/cakephp-shim/src/Model/Behavior/NullableBehavior.php, line 44
  | Shim\Model\Behavior\NullableBehavior::beforeMarshal() - ROOT/vendor/dereuromark/cakephp-shim/src/Model/Behavior/NullableBehavior.php, line 24
  | Cake\Event\EventManager::_callListener() - CORE/src/Event/EventManager.php, line 353
  | Cake\Event\EventManager::dispatch() - CORE/src/Event/EventManager.php, line 330
  | Cake\ORM\Table::dispatchEvent() - CORE/src/Event/EventDispatcherTrait.php, line 114
  | Cake\ORM\Marshaller::_prepareDataAndOptions() - CORE/src/ORM/Marshaller.php, line 291
  | Cake\ORM\Marshaller::one() - CORE/src/ORM/Marshaller.php, line 170
  | Cake\ORM\Table::newEntity() - CORE/src/ORM/Table.php, line 2628
  | App\Controller\AppController::_circumventEventManager() - ROOT/vendor/muffin/footprint/src/Auth/FootprintAwareTrait.php, line 138
  | App\Controller\AppController::_getUserInstanceFromArray() - ROOT/vendor/muffin/footprint/src/Auth/FootprintAwareTrait.php, line 154
  | App\Controller\AppController::_getUserInstance() - ROOT/vendor/muffin/footprint/src/Auth/FootprintAwareTrait.php, line 124
  | App\Controller\AppController::_setCurrentUser() - ROOT/vendor/muffin/footprint/src/Auth/FootprintAwareTrait.php, line 107
  | App\Controller\AppController::_getCurrentUser() - ROOT/vendor/muffin/footprint/src/Auth/FootprintAwareTrait.php, line 78
  | App\Controller\AppController::footprint() - ROOT/vendor/muffin/footprint/src/Auth/FootprintAwareTrait.php, line 57
  | Cake\Event\EventManager::_callListener() - CORE/src/Event/EventManager.php, line 353
  | Cake\Event\EventManager::dispatch() - CORE/src/Event/EventManager.php, line 330
  | Cake\ORM\Table::dispatchEvent() - CORE/src/Event/EventDispatcherTrait.php, line 114
  | Cake\ORM\Table::__construct() - CORE/src/ORM/Table.php, line 300
  | Cake\ORM\Locator\TableLocator::_create() - CORE/src/ORM/Locator/TableLocator.php, line 257
  | Cake\ORM\Locator\TableLocator::get() - CORE/src/ORM/Locator/TableLocator.php, line 224
  | Cake\Controller\Controller::loadModel() - CORE/src/Datasource/ModelAwareTrait.php, line 112
  | Cake\Controller\Controller::__get() - CORE/src/Controller/Controller.php, line 374
  | App\Controller\Admin\EnergyAuditsController::view() - APP/Controller/Admin/EnergyAuditsController.php, line 123
  | Cake\Controller\Controller::invokeAction() - CORE/src/Controller/Controller.php, line 609
  | Cake\Http\ActionDispatcher::_invoke() - CORE/src/Http/ActionDispatcher.php, line 120
  | Cake\Http\ActionDispatcher::dispatch() - CORE/src/Http/ActionDispatcher.php, line 94
  | Cake\Http\BaseApplication::__invoke() - CORE/src/Http/BaseApplication.php, line 235
  | Cake\Http\Runner::__invoke() - CORE/src/Http/Runner.php, line 65
  | Cake\Routing\Middleware\RoutingMiddleware::__invoke() - CORE/src/Routing/Middleware/RoutingMiddleware.php, line 156
  | Cake\Http\Runner::__invoke() - CORE/src/Http/Runner.php, line 65
  | Cake\Routing\Middleware\AssetMiddleware::__invoke() - CORE/src/Routing/Middleware/AssetMiddleware.php, line 88
  | Cake\Http\Runner::__invoke() - CORE/src/Http/Runner.php, line 65
  | Cake\Error\Middleware\ErrorHandlerMiddleware::__invoke() - CORE/src/Error/Middleware/ErrorHandlerMiddleware.php, line 98
  | Cake\Http\Runner::__invoke() - CORE/src/Http/Runner.php, line 65
  | DebugKit\Middleware\DebugKitMiddleware::__invoke() - ROOT/vendor/cakephp/debug_kit/src/Middleware/DebugKitMiddleware.php, line 53
  | Cake\Http\Runner::__invoke() - CORE/src/Http/Runner.php, line 65
  | Cake\Http\Runner::run() - CORE/src/Http/Runner.php, line 51
  | Cake\Http\Server::run() - CORE/src/Http/Server.php, line 98
  | [main] - ROOT/webroot/index.php, line 40
```

Reply:
What is the data that gets passed into the beforeMarshal() callback?
As you might have seen, that one already doesnt allow null here

    beforeMarshal(Event $event, ArrayObject $data, ArrayObject $options)

this is also as documented only ArrayObject as type.

This makes me think the bug is in that other plugin, passing invalid data into callbacks.
What do you think?
How does your PR solve this, if the callback before is already not accepting null?

Reply:
Ok so this is gonna be long but I think the error is in the Shim Plugin. Attached is my modified NullableBehavior and the output 

```
/vendor/dereuromark/cakephp-shim/src/Model/Behavior/NullableBehavior.php (line 25)
object(Cake\Event\Event) {
	result => null
	[protected] _name => 'Model.beforeMarshal'
	[protected] _subject => object(App\Model\Table\UsersTable) {

		'registryAlias' => 'Users',
		'table' => 'users',
		'alias' => 'Users',
		'entityClass' => 'App\Model\Entity\User',
		'associations' => [
			(int) 0 => 'tenants',
			(int) 1 => 'createdusers',
			(int) 2 => 'modifiedusers',
			(int) 3 => 'socialaccounts'
		],
		'behaviors' => [
			(int) 0 => 'Timestamp',
			(int) 1 => 'Nullable',
			(int) 2 => 'Register',
			(int) 3 => 'Password',
			(int) 4 => 'Social',
			(int) 5 => 'LinkSocial'
		],
		'defaultConnection' => 'default',
		'connectionName' => 'default'
	
	}
	[protected] _data => [
		'data' => object(ArrayObject) {
			id => (int) 1
			tenant_id => null
			username => null
			first_name => 'Admin'
			last_name => 'User'
			email => 'admin@energyauditapp.com'
			phone => null
			timezone => 'America/Chicago'
			tos_date => object(Cake\I18n\FrozenTime) {

				'time' => '2018-11-01T00:00:00-05:00',
				'timezone' => 'America/Chicago',
				'fixedNowTime' => false
			
			}
			activation_date => object(Cake\I18n\FrozenTime) {

				'time' => '2018-11-01T00:00:00-05:00',
				'timezone' => 'America/Chicago',
				'fixedNowTime' => false
			
			}
			active => true
			is_superuser => true
			role => 'admin'
			secret => null
			secret_verified => null
			created_user_id => null
			modified_user_id => (int) 1
			created => object(Cake\I18n\FrozenTime) {

				'time' => '2018-11-01T00:00:00-05:00',
				'timezone' => 'America/Chicago',
				'fixedNowTime' => false
			
			}
			modified => object(Cake\I18n\FrozenTime) {

				'time' => '2018-11-01T00:00:00-05:00',
				'timezone' => 'America/Chicago',
				'fixedNowTime' => false
			
			}
			energy_audit_count => (int) 0
			tenant => null
			fullname => 'Admin User'
		},
		'options' => object(ArrayObject) {
			accessibleFields => [
				'*' => true
			]
			validate => false
			associated => [
				(int) 0 => 'tenants',
				(int) 1 => 'createdusers',
				(int) 2 => 'modifiedusers',
				(int) 3 => 'socialaccounts'
			]
		}
	]
	[protected] _stopped => false
}
/vendor/dereuromark/cakephp-shim/src/Model/Behavior/NullableBehavior.php (line 55)
null
/vendor/dereuromark/cakephp-shim/src/Model/Behavior/NullableBehavior.php (line 56)
'tenant'
/vendor/dereuromark/cakephp-shim/src/Model/Behavior/NullableBehavior.php (line 57)
[
	'tenant' => 'Tenants',
	'created_user' => 'CreatedUsers',
	'modified_user' => 'ModifiedUsers',
	'social_accounts' => 'SocialAccounts'
]
/vendor/dereuromark/cakephp-shim/src/Model/Behavior/NullableBehavior.php (line 46)
object(App\Model\Table\TenantsTable) {

	'registryAlias' => 'Tenants',
	'table' => 'tenants',
	'alias' => 'Tenants',
	'entityClass' => 'App\Model\Entity\Tenant',
	'associations' => [
		(int) 0 => 'states',
		(int) 1 => 'energyfixtureoptions',
		(int) 2 => 'energyfixturereplacements',
		(int) 3 => 'energyfixtureretrofits',
		(int) 4 => 'users'
	],
	'behaviors' => [
		(int) 0 => 'Timestamp',
		(int) 1 => 'Nullable'
	],
	'defaultConnection' => 'default',
	'connectionName' => 'default'

}
/vendor/dereuromark/cakephp-shim/src/Model/Behavior/NullableBehavior.php (line 47)
null
Shim\Model\Behavior\NullableBehavior::_process() - ROOT/vendor/dereuromark/cakephp-shim/src/Model/Behavior/NullableBehavior.php, line 49
Shim\Model\Behavior\NullableBehavior::_process() - ROOT/vendor/dereuromark/cakephp-shim/src/Model/Behavior/NullableBehavior.php, line 58
Shim\Model\Behavior\NullableBehavior::beforeMarshal() - ROOT/vendor/dereuromark/cakephp-shim/src/Model/Behavior/NullableBehavior.php, line 31
Cake\Event\EventManager::_callListener() - CORE/src/Event/EventManager.php, line 353
Cake\Event\EventManager::dispatch() - CORE/src/Event/EventManager.php, line 330
Cake\ORM\Table::dispatchEvent() - CORE/src/Event/EventDispatcherTrait.php, line 114
Cake\ORM\Marshaller::_prepareDataAndOptions() - CORE/src/ORM/Marshaller.php, line 291
Cake\ORM\Marshaller::one() - CORE/src/ORM/Marshaller.php, line 170
Cake\ORM\Table::newEntity() - CORE/src/ORM/Table.php, line 2628
App\Controller\AppController::_circumventEventManager() - ROOT/vendor/muffin/footprint/src/Auth/FootprintAwareTrait.php, line 138
App\Controller\AppController::_getUserInstanceFromArray() - ROOT/vendor/muffin/footprint/src/Auth/FootprintAwareTrait.php, line 154
App\Controller\AppController::_getUserInstance() - ROOT/vendor/muffin/footprint/src/Auth/FootprintAwareTrait.php, line 124
App\Controller\AppController::_setCurrentUser() - ROOT/vendor/muffin/footprint/src/Auth/FootprintAwareTrait.php, line 107
App\Controller\AppController::_getCurrentUser() - ROOT/vendor/muffin/footprint/src/Auth/FootprintAwareTrait.php, line 78
App\Controller\AppController::footprint() - ROOT/vendor/muffin/footprint/src/Auth/FootprintAwareTrait.php, line 61
Cake\Event\EventManager::_callListener() - CORE/src/Event/EventManager.php, line 353
Cake\Event\EventManager::dispatch() - CORE/src/Event/EventManager.php, line 330
Cake\Controller\Component\AuthComponent::dispatchEvent() - CORE/src/Event/EventDispatcherTrait.php, line 114
Cake\Controller\Component\AuthComponent::identify() - CORE/src/Controller/Component/AuthComponent.php, line 827
CakeDC\Users\Controller\UsersController::login() - ROOT/vendor/cakedc/users/src/Controller/Traits/LoginTrait.php, line 182
Cake\Controller\Controller::invokeAction() - CORE/src/Controller/Controller.php, line 609
Cake\Http\ActionDispatcher::_invoke() - CORE/src/Http/ActionDispatcher.php, line 120
Cake\Http\ActionDispatcher::dispatch() - CORE/src/Http/ActionDispatcher.php, line 94
Cake\Http\BaseApplication::__invoke() - CORE/src/Http/BaseApplication.php, line 235
Cake\Http\Runner::__invoke() - CORE/src/Http/Runner.php, line 65
Cake\Routing\Middleware\RoutingMiddleware::__invoke() - CORE/src/Routing/Middleware/RoutingMiddleware.php, line 156
Cake\Http\Runner::__invoke() - CORE/src/Http/Runner.php, line 65
Cake\Routing\Middleware\AssetMiddleware::__invoke() - CORE/src/Routing/Middleware/AssetMiddleware.php, line 88
Cake\Http\Runner::__invoke() - CORE/src/Http/Runner.php, line 65
Cake\Error\Middleware\ErrorHandlerMiddleware::__invoke() - CORE/src/Error/Middleware/ErrorHandlerMiddleware.php, line 98
Cake\Http\Runner::__invoke() - CORE/src/Http/Runner.php, line 65
DebugKit\Middleware\DebugKitMiddleware::__invoke() - ROOT/vendor/cakephp/debug_kit/src/Middleware/DebugKitMiddleware.php, line 53
Cake\Http\Runner::__invoke() - CORE/src/Http/Runner.php, line 65
Cake\Http\Runner::run() - CORE/src/Http/Runner.php, line 51
Cake\Http\Server::run() - CORE/src/Http/Server.php, line 98
[main] - ROOT/webroot/index.php, line 40
```

```
<?php
namespace Shim\Model\Behavior;

use ArrayObject;
use Cake\Event\Event;
use Cake\ORM\Behavior;
use Cake\ORM\Table;
use Cake\Utility\Hash;

/**
 * A behavior to assert data consistency regarding empty values.
 * It will check the schema and depending on nullable type or not transform empty string input
 * (from forms for example) into `null`.
 */
class NullableBehavior extends Behavior {

	/**
	 * @param \Cake\Event\Event $event
	 * @param \ArrayObject $data
	 * @param \ArrayObject $options
	 * @return void
	 */
	public function beforeMarshal(Event $event, ArrayObject $data, ArrayObject $options)
	{
		debug($event);
		if ($data === null) {
			debug($event);
			debug($data);
			die;
		}
		$this->_process($data, $this->_table);
	}

	/**
	 * @param \ArrayObject $data
	 * @param \Cake\ORM\Table $table
	 * @return \ArrayObject
	 */
	protected function _process($data, Table $table) {
		$associations = [];
		/** @var \Cake\ORM\Association $association */
		foreach ($table->associations() as $association) {
			$associations[$association->getProperty()] = $association->getName();
		}
		if ($data === null) {
			debug($table);
			debug($data);
			echo '<pre>';
			stackTrace();
			echo '</pre>';die;
			return $data;
		}
		foreach ($data as $key => $value) {
			if (array_key_exists($key, $associations)) {
				debug($data[$key]);
				debug($key);
				debug($associations);
				$data[$key] = $this->_process($data[$key], $table->getAssociation($associations[$key])->getTarget());
				continue;
			}
			$nullable = Hash::get((array)$table->getSchema()->getColumn($key), 'null');
			if ($nullable !== true) {
				continue;
			}
			if ($value !== '') {
				continue;
			}

			$default = Hash::get((array)$table->getSchema()->getColumn($key), 'default');
			$data[$key] = $default;
		}

		return $data;
	}

}
```